Back port HTTP/2 codec from master to 4.1

NiteshKant opened the pull request:
##### Motivation

HTTP/2 codec was implemented in master branch.
Since, master is not yet stable and will be some time before it gets released, backporting it to 4.1, enables people to use the codec with a stable netty version.

##### Modification:

The code has been copied from master branch as is, with minor modifications to suit the `ChannelHandler` API in 4.x.
Apart from that change, there are two backward incompatible API changes included, namely,

- Added an abstract method:

  ```java
public abstract Map.Entry<CharSequence, CharSequence> forEachEntry(EntryVisitor<CharSequence> visitor)
            throws Exception;
```

   to `HttpHeaders` and implemented the same in `DefaultHttpHeaders` as a delegate to the internal `TextHeader` instance.

- Added a method:

```java
FullHttpMessage copy(ByteBuf newContent);
```
  in `FullHttpMessage` with the implementations copied from relevant places in the master branch.

##### Result:

HTTP/2 codec can be used with netty 4.1

netkins commented:
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

normanmaurer commented:
@netkins accept

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaders.java:
```diff
@@ -1550,6 +1556,27 @@ public HttpHeaders set(HttpHeaders headers) {
+            throw new NullPointerException("headers");
+        }
+
+        if (headers.isEmpty()) {
```
Consider staying consistent with [DefaultHeaders.setAll(..)](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java#L743).  This involves using the `forEachEntry` construct.

NiteshKant replied:
Fixed. I have also updated `HttpHeaders.set()` to use `forEachEntry()` as is done by `DefaultHeaders`

I also added a set of set/addShort() methods matching what we have in `DefaultHeaders`. `setShort()` was required by `InboundHttp2ToHttpPriorityAdapter`

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaders.java:
```diff
@@ -1550,6 +1556,27 @@ public HttpHeaders set(HttpHeaders headers) {
+     * @return {@code this}
+     */
+    public HttpHeaders setAll(HttpHeaders headers) {
+        if (headers == null) {
```
use `checkNotNull`.

NiteshKant replied:
Done

Scottmitch commented on pom.xml:
```diff
@@ -555,6 +556,11 @@
+        <groupId>com.twitter</groupId>
+        <artifactId>hpack</artifactId>
+        <version>0.9.1</version>
+      </dependency>
```
why not `0.10.0`?

Scottmitch commented on pom.xml:
```diff
@@ -555,6 +556,11 @@
+      <dependency>
+        <groupId>com.twitter</groupId>
+        <artifactId>hpack</artifactId>
+        <version>0.9.1</version>
```
why not 0.10.0?

Scottmitch replied:
Can you confirm again from what commit you have done the backport from?

NiteshKant replied:
My bad, I did not update the pom.xml to [commit hash](https://github.com/netty/netty/commit/14039623137f2bfdaec9a7228e72fe94c663ded5) (which is the commit I have backported from).
Fixed it now to 0.10.0.

Scottmitch commented:
I have done a quick review and only focused on the areas that looked new (HttpHeaders interface related code).

It sounds like you didn't have to make any modifications to http2 codec code?  How did you resolve the unit test failing issue you mentioned [here](https://github.com/netty/netty/issues/3259#issuecomment-70588137)?  Are there any other areas you would like us to focus on in review (content is rather large in this PR :) )?

NiteshKant commented:
Thanks @Scottmitch for the review!

> I have done a quick review and only focused on the areas that looked new

I added a few add/setShort() methods to `HttpHeaders`, which you may want to look.

> It sounds like you didn't have to make any modifications to http2 codec code?

I had to do the following minor modifications:

- `Http2ConnectionHandler` had to implement `ChannelOutboundHandler` to get a callback for `close()`. Other methods just delegate to corresponding methods in `ChannelHandlerContext`.
- `Http2ServerUpgradeCodec.prepareUpgradeResponse()` used 
```java
List<CharSequence> upgradeHeaders = upgradeRequest.headers().getAll(HTTP_UPGRADE_SETTINGS_HEADER);
```
which had to be changed to:
```java
List<String> upgradeHeaders = upgradeRequest.headers().getAll(HTTP_UPGRADE_SETTINGS_HEADER);
```
(Type of the returned list changed)

- `HttpUtil.toHttp2Headers()` was using:

```java
String value = inHeaders.getAndConvert(HttpHeaderNames.HOST);
```

which I changed to:

```java
String value = inHeaders.get(HttpHeaderNames.HOST);
```

as `HttpHeaders` does not have a `getAndConvert()` function and did not appear to me that there is any semantic change in the two methods.

- `InboundHttp2ToHttpAdapterTest.HttpResponseDelegator` and `InboundHttp2ToHttpAdapterTest.HttpSettingsDelegator` `messageReceived` methods were renamed to `channelRead0` as per the `SimpleChannelInboundHandler` API.

> How did you resolve the unit test failing issue you mentioned here? 

The test wasn't flaky, I had done a code change to accommodate the missing `FullHttpMessage.copy(ByteBuf)` method, which I reverted after adding the same. That fixed the test.

> Are there any other areas you would like us to focus on in review (content is rather large in this PR :) )

Yeah the content is large. I think if you can review the http codec changes, it will be great. I have outlined the http2 codec changes above, so you can review them too.

Thanks!

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaders.java:
```diff
@@ -131,11 +154,32 @@ public HttpHeaders clear() {
         }
     };
 
+    private final EntryVisitor<CharSequence> setAllVisitor = new EntryVisitor<CharSequence>() {
```
Thanks for being consistent :).  This has lead me to re-review the DefaultHeaders implementation and realize that this member is a good candidate for lazy initialization.  See PR https://github.com/netty/netty/pull/3342.  If you feel this is a good improvement then lets follow the lazy initialization approach.

NiteshKant replied:
Yep I saw that PR & the change makes sense. Will make the change here too. 
Quick question:
Not that it matters in this case, but HttpHeaders isn't threadsafe, correct?

Scottmitch replied:
> Not that it matters in this case, but HttpHeaders isn't threadsafe, correct?

Correct. [DefaultHeaders](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java) is the base class for most header functionality and it is not thread safe. The [Headers](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java) currently does not guarantee thread safety.

Scottmitch commented:
> Http2ConnectionHandler had to implement ChannelOutboundHandler to get a callback for close(). Other methods just delegate to corresponding methods in ChannelHandlerContext.

Yes due to some channel interface restructuring.  I think it should be OK (especially if all the methods are still getting invoked :) ).  @normanmaurer - Thoughts?
In [5.x](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L46) the class is:
```java
public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http2LifecycleManager
```

in [this PR](https://github.com/netty/netty/pull/3339/files#diff-ce33d9374618cacba50c2b6020df1c86R48) (4.1) the class is:
```java
public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http2LifecycleManager, ChannelOutboundHandler
```

> HttpHeaders does not have a getAndConvert() function and did not appear to me that there is any semantic change in the two methods.

Good pickup. The two different methods (`get()` in HttpHeaders for 4.x, and `getAndConvert()` in 5.x) should be equivalent. The `AndConvert` semantics are due to the generic type being `CharSequence` and a conversion may be required to get to `String`.  Prior to headers restructure the conversion operation was often not needed and not clear that it was being done.

> The test wasn't flaky, I had done a code change to accommodate the missing FullHttpMessage.copy(ByteBuf) method, which I reverted after adding the same. That fixed the test.

Yay unit tests :)

All the other changes you describe seem OK.

@nmittler - Hey buddy, how have you been :)  Do you have any thoughts on this?

nmittler commented:
@NiteshKant The changes LGTM! Thanks for doing this! :)

NiteshKant commented:
Thanks @nmittler for the review. @Scottmitch I will do changes outlined in PR #3342 here and then should I assume this is ready to merge?

normanmaurer commented:
@NiteshKant I would say so... I would like to keep the PR as much in line with master as possible and do optimisations that apply to both branches later. Just to keep it simple and not miss anything later.

Scottmitch commented:
@NiteshKant - Also look at https://github.com/netty/netty/pull/3343.   I just removed the member variables all together.

After these changes yes I think it is good to merge.

Scottmitch commented:
>  I would like to keep the PR as much in line with master as possible and do optimisations that apply to both branches late

@normanmaurer - Agreed.  I think the changes we are discussing involve the differences in implementations between the HttpHeaders interface.  These differences are not directly ported between branches (due to backward compatibility issues), and so I think doing them here is the right thing to do (because they do not conflict with other branches).  Does this make sense?

normanmaurer commented:
@Scottmitch yes

NiteshKant commented:
@Scottmitch I included changes from #3342 #3343 into this PR. Lemme know if this looks fine.

PS: I am not sure why the CI build stopped building this PR.

Scottmitch commented:
> @Scottmitch I included changes from #3342 #3343 into this PR. Lemme know if this looks fine.

Looks good.  Thanks!

> PS: I am not sure why the CI build stopped building this PR.

You have a few jobs queued up. See http://clinker.netty.io/jenkins/view/Pulls/job/netty-pullreqs/.  Oracle jdk8 for branch 4.0 is building now see http://clinker.netty.io/jenkins/.  I'm not sure how much parallelism is currently supported on the build server.

normanmaurer commented:
Just a reminder to also apply the PR https://github.com/netty/netty/pull/3348 after this is merged

Scottmitch commented:
@NiteshKant - Thanks a bunch!

Cherry-picked into 4.1 (2d24e1f27d00d13401d543a04beb5b76102dfd7d).

Scottmitch commented:
@nmittler - Just a reminder now we have to backport fixes :)

nmittler commented:
@Scottmitch yup ... thanks for the heads up! :)

NiteshKant commented:
@Scottmitch My pleasure :) Is there a plan on when the next 4.1 release will be available?

BTW, If you guys need help in backporting any fixes, lemme know.

Scottmitch commented:
> Is there a plan on when the next 4.1 release will be available?

@trustin is the man to ask for that.  I can refer you to http://netty.io/news/2014/12/31/4-0-25-Final.html which says this month.  We also have https://github.com/netty/netty/issues/2933 out standing to track discussion related to 5.0 release.

> BTW, If you guys need help in backporting any fixes, lemme know.

@NiteshKant - Will do.  However I think we are in a good place now since most code is identical.

normanmaurer commented:
@NiteshKant @Scottmitch yeah we should definitely try to release over the next weeks.
